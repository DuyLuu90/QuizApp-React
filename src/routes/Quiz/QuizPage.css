.js-quiz, .js-question, .alertMessage{
    width: 90%;
    max-width: 600px;
    margin: 15px auto;
    padding: 15px;
}
.js-quiz  {
    border-style: solid;
    border-width: thin;  
}
.js-quiz h1{
    color:  rgb(11, 11, 179);
    margin-bottom: 15px;
    font-size: 25px;
}
.js-quiz li{
    text-align: left;
    padding: 10px;
    background-color: rgb(11, 11, 179); color: white;
    border-radius: 10px;
    margin: 10px auto;
}
.js-quiz li:hover {
    background-color: orangered;
    color: black;
}
.js-quiz button {
    padding: 10px;
    background-color: red;
    color: white;
    margin-top: 15px;
}
.control {
    display: flex;
    justify-content: space-between;
    margin: 10px auto;
}
.control a {
    padding: 10px;
    background-color: blue;
    color: white;
}
.control div{
    padding: 10px;
    font-weight: bolder;
    color: black;
    background-color: rgb(230, 156, 20)
}
#quiz-info{
    width: 250px;
}
#quiz-final{
    width: 180px;
}