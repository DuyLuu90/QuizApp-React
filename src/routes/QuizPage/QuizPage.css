.js-quiz, .js-question, .alertMessage{
    width: 90%;
    max-width: 600px;
    margin: 15px auto;
    padding: 15px;
}
.js-quiz  {
    border-style: solid;
    border-width: thin;  
}
.js-quiz h1{
    color:  rgb(6, 6, 124);
    margin-bottom: 15px;
    font-size: 25px;
}
.js-quiz li{
    text-align: left;
    padding: 10px;
    background-color: rgb(57,75,87); color: white;
    border-radius: 10px;
    margin: 10px auto;
}
.js-quiz li:hover {
    background-color: orangered
}
#final{
    margin-top: 100px;
    color: rgb(34, 21, 21);
    background-color:  rgba(245, 245, 245, 0.89)
}
.control {
    display: flex;
    justify-content: space-between;
    margin: 10px auto;
}
.control a {
    padding: 10px;
}
.control div{
    padding: 10px;
    font-weight: bolder;
    color: black;
    background-color: rgb(230, 156, 20)
}
#quiz-info{
    width: 250px;
}
#quiz-final{
    width: 180px;
}