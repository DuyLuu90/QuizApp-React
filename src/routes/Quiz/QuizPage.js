import React, {Component} from 'react'
import {Link} from 'react-router-dom'
import './QuizPage.css'
import Question from '../../component/question/question'
import {quizBank} from '../../assets/QuizBank'
import helpers from '../../services/helpers'

export default class QuizPage extends Component {
    static defaultProps = {

    }
    state={
        displayInstruction: true,
        displayFinal: false,
        quizBank:[],
        currentQuestion:0,
        score:0
    }
    componentDidMount(){
        helpers.shuffleArray(quizBank)
        const array= quizBank.slice(0,10)
        this.setState({quizBank: array})
    }
    updateQuestion=()=>{
        const {currentQuestion,quizBank}= this.state
        console.log(quizBank.length)
        if (currentQuestion+1 === quizBank.length){
            this.setState({displayFinal: true})
        }
        else {
            this.setState({currentQuestion: currentQuestion+1})
        } 
    
    }
    updateScore=()=>{
        this.setState({score: this.state.score+1})
    }
    playNow=()=>{
        this.setState({displayInstruction: false})
    }
    renderInstructions(){
        return(
            <div className='js-quiz'>
                <h1>INSTRUCTIONS</h1>
                <ul>
                    <li>A round of trivia contains 10 multiple-choice questions</li>
                    <li>After you submit your answer for a question,the correct answer will be revealed, and your score will be updated</li>
                    <li>You have to submit the answer in order to move to the next question. </li>
                    <li>At the end of the round, you will see your final score. That's very simple. Hope you'll have some fun and good luck! </li>
                </ul>
                <button type='button' onClick={this.playNow}>PLAY NOW</button>
            </div>
        )
    }
    /*
    renderQuestions(){
        const questions= this.state.quizBank.map((question,i)=>
            <Question key={i}item={question} 
                updateQuestion={this.updateQuestion}
                updateScore={this.updateScore}
            />)
        return questions;
    }*/
    renderQuestion(){
        const {currentQuestion,score,quizBank}= this.state
        const question= quizBank[currentQuestion]
        console.log(question)
        return (
            <div>
                <div className='control' id='quiz-info'> 
                    <div>
                        Question: {currentQuestion+1}/10
                    </div>
                    <div>
                        Score: {score}/10
                    </div>
                </div>
                <Question item={question} 
                updateQuestion={this.updateQuestion}
                updateScore={this.updateScore}/>
            </div>
        )
    }
    renderFinalScreen(){
        const {score}= this.state
        return(
            <div className='js-quiz'>
                <div>
                    <h1>THANK YOU</h1>
                    <p> You have completed your quiz. Your score is <strong>{score}/10</strong>. What would you like to do next? </p>
                </div>
                <div className='control' id='quiz-final'>
                    <Link to='/' aria-label='home-page'>EXIT</Link>
                    <Link to='/quiz' aria-label='quiz-page' onClick={()=>window.location.reload(false)}>RETAKE</Link>
                </div>
            </div>
        )
    }

    render(){
        const {displayInstruction,displayFinal}= this.state
        const instruction= this.renderInstructions()
        const questions= this.renderQuestion() 
        const final= this.renderFinalScreen()
        /*
        const html= (displayInstruction)? instruction
                    : (displayFinal)? final
                    : questions*/
        return(
            <div>
                {displayInstruction && instruction}
                {!displayFinal && !displayInstruction && questions}
                {displayFinal && final}
            </div> 
        )
    }
}