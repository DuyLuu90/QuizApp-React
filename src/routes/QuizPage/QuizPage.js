import React, {Component} from 'react'
import {Prompt} from 'react-router-dom'
import './QuizPage.css'
import Question from '../../component/question/question'
import {quizBank} from '../../assets/QuizBank'
import helpers from '../../services/helpers'

export default class QuizPage extends Component {
    static defaultProps = {
        history: {}
    }
    state={
        displayInstruction: true,
        displayFinal: false,
        quizBank:[],
        currentQuestion:0,
        score:0,
        playNow: false,
        time:0,
    }
    componentDidMount(){
        helpers.shuffleArray(quizBank)
        const array= quizBank.slice(0,10)
        this.setState({quizBank: array})
    }

    startClock=()=>{
        this.myInterval= setInterval(()=>{
            this.setState({time: this.state.time+1})
        },1000)
    }
    stopClock=()=>{
        clearInterval(this.myInterval)
    }
    updateQuestion=()=>{
        const {currentQuestion,quizBank}= this.state
        if (currentQuestion+1 === quizBank.length){
            this.stopClock()
            this.setState({
                playNow: false,
                displayFinal: true})
        }
        else {
            this.setState({currentQuestion: currentQuestion+1})
        } 
    }
    updateScore=()=>this.setState({score: this.state.score+1})
    
    playNow=()=>{
        this.setState({
            playNow: true,
            displayInstruction: false})
        this.startClock()
    }

    renderInstructions(){
        return(
            <div className='js-quiz'>
                <h1>INSTRUCTIONS</h1>
                <ul>
                    <li>A round of trivia contains 10 multiple-choice questions</li>
                    <li>After submitting your answer,the correct answer will be revealed, and your score will be updated</li>
                    <li>Submit the answer in order to move to the next question. </li>
                    <li>At the end of the round, you will see your final score. That's very simple. Have fun and good luck! </li>
                </ul>
                <button type='button' onClick={this.playNow}>PLAY NOW</button>
            </div>
        )
    }
    renderQuestion(){
        const {currentQuestion,score,quizBank,time}= this.state
        const question= quizBank[currentQuestion]
        const clock= new Date(time*1000).toISOString().substr(11,8)
        return (
            <div>
                <div className='timer'>
                    {clock}
                </div>
                <div className='control' id='quiz-info'> 
                    <div>
                        Question: {currentQuestion+1}/10
                    </div>
                    <div>
                        Score: {score}/10
                    </div>
                </div>
                <Question 
                    item={question} 
                    updateQuestion={this.updateQuestion}
                    updateScore={this.updateScore}
                />
            </div>
        )
    }
    renderFinalScreen(){
        const {score, time}= this.state
        const {history}= this.props
        const clock= new Date(time*1000).toISOString().substr(11,8).split(':')
        const hours= Number(clock[0])
        const mins= Number(clock[1])
        const secs = Number(clock[2])
        return(
            <div className='js-quiz' id='final'>
                <div>
                    <h1>THANK YOU</h1>
                    <p> You completed your quiz in 
                        {hours>0 && <strong> {hours} hours </strong>}
                        {mins>0 && <strong className='red'> {mins} minutes</strong>}
                        {secs>0 && <strong className='green'> {secs} seconds</strong>}
                        . Your score is <strong>{score}/10</strong>. What would you like to do next? </p>
                </div>
                <div className='control' id='quiz-final'>    
                    <button type='button' onClick={()=>history.push('/')}>EXIT</button>
                    <button type='button' onClick={()=>window.location.reload()}>RETAKE</button> 
                </div>
            </div>
        )
    }
    
    render(){
        const {displayInstruction,displayFinal}= this.state
        const instruction= this.renderInstructions()
        const questions= this.renderQuestion() 
        const final= this.renderFinalScreen()
        const html=   displayInstruction  ? instruction
                    : displayFinal        ? final
                    : questions
        
        return (
            <div>
                <Prompt 
                    when={this.state.playNow}
                    message={JSON.stringify({
                        content:'You have not finished your quiz. Are you sure you want to leave?'
                    })}/>
                {html}
            </div>
        )
    }
}
