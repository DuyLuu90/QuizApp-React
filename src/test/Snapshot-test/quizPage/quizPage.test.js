import { faItalic } from '@fortawesome/free-solid-svg-icons';
import React from 'react';
import {BrowserRouter} from 'react-router-dom'
import renderer from 'react-test-renderer'

import QuizPage from '../../../routes/QuizPage/QuizPage';

it.only(`Snapshot test on Quiz Page to prevent unexpected change`,()=>{
    const myRenderedElement= renderer.create(
        <BrowserRouter>
            <QuizPage />
        </BrowserRouter>).toJSON();
    expect(myRenderedElement).toMatchSnapshot();
})