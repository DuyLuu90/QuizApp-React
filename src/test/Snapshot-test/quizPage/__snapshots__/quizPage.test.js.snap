// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test on Quiz Page to prevent unexpected change 1`] = `
<div
  className="js-quiz"
>
  <h1>
    INSTRUCTIONS
  </h1>
  <ul>
    <li>
      A round of trivia contains 10 multiple-choice questions
    </li>
    <li>
      After submitting your answer,the correct answer will be revealed, and your score will be updated
    </li>
    <li>
      Submit the answer in order to move to the next question. 
    </li>
    <li>
      At the end of the round, you will see your final score. That's very simple. Have fun and good luck! 
    </li>
  </ul>
  <button
    onClick={[Function]}
    type="button"
  >
    PLAY NOW
  </button>
</div>
`;
